src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs

```diff
@@ -57,8 +66,27 @@ async Task<CompletableSynchronizedStorageSession> AdaptOrOpenNewSynchronizedStor
                    ?? await synchronizedStorage.OpenSession(contextBag).ConfigureAwait(false);
         }
 
+        static void LogHandlersInvocation(IIncomingLogicalMessageContext context, List<MessageHandler> handlersToInvoke)
+        {
+            var builder = new StringBuilder();
+            builder.AppendLine($"Processing message type: {context.Message.MessageType}");
+            builder.AppendLine($"Headers: {Environment.NewLine}{string.Join(", ", context.Headers.Select(h => $"{h.Key}:{h.Value}").ToArray())}");
```

Comment:
I know this is only used when runnig debug logging, but our general guideline is to not use LINQ in any hot path.

Reply:
I followed the approach used in the outgoing connector.
If the requirement to change it then should it be applied in both places rather than one?
Also, out if curiosity, as you've mentioned this will only be executed when debug is turned on. Doesn't that imply impacted performance because of the debug nature?

Reply:
> If the requirement to change it then should it be applied in both places rather than one?

the guidelines primarily affect newly written code. If you want to improve the code in the outgoing connector, that sure would be a welcomed change 👍 .

> as you've mentioned this will only be executed when debug is turned on. Doesn't that imply impacted performance because of the debug nature?

the LINQ code is still fairly easy to avoid and we should continue to avoid unnecessary allocations  in the hot path, regardless of the log level (which comes with some unavoidable overhead).

Reply:
I'm not convinced that removing this specific LINQ code adds significant value. While it's on the hot path, it's _only_ executed when debug is turned on. Considering that, I am not convinced it's necessary, but will follow change it as per @Particular/nservicebus-maintainers request... If you do think LINQ should stay in place, will revert last commit.